strawberryfields/api/devicespec.py

```diff
@@ -0,0 +1,141 @@
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains a class that represents the specifications of
+a device available via the API.
+"""
+from collections.abc import Sequence
+
+import strawberryfields as sf
+import blackbird
+
+from strawberryfields.circuitspecs import Ranges
+
+
+class DeviceSpec:
+    """The specifications for a specific hardware device.
+
+    Args:
+        target (str): name of the target hardware device
+        device (dict): dictionary representing the raw device specification.
+            This dictionary should contain the following key-value pairs:
+
+            - layout (str): string containing the Blackbird circuit layout
+            - modes (int): number of modes supported by the target
+            - compiler (list): list of supported compilers
+            - gate_parameters (dict): parameters for the circuit gates
+
+        connection (strawberryfields.api.Connection): connection over which the
+            job is managed
+    """
+
+    def __init__(self, target, device, connection):
+        self._target = target
+        self._connection = connection
+        self._spec = device
+
+    @property
+    def target(self):
+        """str: The name of the target hardware device."""
+        return self._target
+
+    @property
+    def layout(self):
+        """str: Returns a string containing the Blackbird circuit layout."""
+        return self._spec["layout"]
+
+    @property
+    def modes(self):
+        """int: Number of modes supported by the device."""
+        return self._spec["modes"]
+
+    @property
+    def compiler(self):
+        """list[str]: A list of strings corresponding to Strawberry Fields compilers supported
+        by the hardware device."""
+        return self._spec["compiler"]
+
+    @property
+    def gate_parameters(self):
+        """dict[str, strawberryfields.circuitspecs.Ranges]: A dictionary of gate parameters
+        and allowed ranges.
+
+        The parameter names correspond to those present in the Blackbird circuit layout.
+
+        **Example**
+
+        >>> spec.gate_parameters
+        {'squeezing_amplitude_0': x=0, x=1, 'phase_0': x=0, 0≤x≤6.283185307179586}
+        """
+        gate_parameters = dict()
+
+        for gate_name, param_ranges in self._spec["gate_parameters"].items():
+            # convert gate parameter allowed ranges to Range objects
+            range_list = [[i] if not isinstance(i, Sequence) else i for i in param_ranges]
+            gate_parameters[gate_name] = Ranges(*range_list, variable_name=gate_name)
```

Comment:
Maybe we should just use `x` as the ranges variable for now (since that's used in the docstring, and it looks a bit cleaner). Then we can update all of this in a later PR?

```suggestion
            gate_parameters[gate_name] = Ranges(*range_list)
```

Reply:
Good idea!